pkg/ipamd/ipamd.go

```diff
@@ -988,8 +997,24 @@ func (c *IPAMContext) nodeIPPoolReconcile(interval time.Duration) {
 
 func (c *IPAMContext) eniIPPoolReconcile(ipPool []string, attachedENI awsutils.ENIMetadata, eni string) {
 	seenIPs := make(map[string]bool)
+	attachedENIIPs := attachedENI.IPv4Addresses
+	isMetadataIPs := true
+	// Here we can't trust attachedENI since the IMDS metadata can be stale. We need to check with EC2 API.
+	// +1 is for the primary IP of the ENI that is not added to the ipPool and not available for pods to use.
+	if 1+len(ipPool) != len(attachedENIIPs) {
+		log.Warnf("Instance metadata does not match data store! ipPool: %v, metadata: %v", ipPool, attachedENIIPs)
+		log.Debugf("We need to check the ENI status by calling the EC2 control plane.")
+		// Call EC2 to verify IPs on this ENI
+		ec2Addresses, err := c.awsClient.GetIPv4sFromEC2(eni)
+		if err != nil {
+			log.Errorf("Failed to fetch ENI IP addresses! Aborting reconcile of ENI %s", eni)
+			return
+		}
+		attachedENIIPs = ec2Addresses
```

Comment:
What happens if the IP is just got allocated? It will be assigned to attachedENIIPs and might get deleted right?

Reply:
If the secondary IP just got added outside of IPAMD, so it's not yet seen in the `ipPool`, but shows up in IMDS metadata. Then we will also call EC2 to verify, then this reconcile function will add it when looping over `attachedENIIPs`:
```
c.dataStore.AddIPv4AddressToStore(eni, strPrivateIPv4)
```